### PR TITLE
Allow redialing to other servers

### PR DESCRIPTION
**Pull request by Elijahrane**
- [x] Proof of concept
- [x] Figure out how to actually register the message properly on client
- [x] Untangle server stuff from entity systems
- [x] Server full UI redirect
- [x] Ghost role menu UI redirect
- [x] Better list filtration (scale to like 90% full max?)
- [x] Better method/format of storing the list

:cl: Rane
- add: You can be redirected to another server when the server is full or from the ghost roles menu.

### CODE REVIEW

**Comment by Elijahrane** [comment]
@ZeroDayDaemon @Vordenburg it works fine in testing; it needs a bit of cleanup though and you guys have more experience with server stuff than me

particularly the URI construction is kind of bad. should I import the whole system from the launcher? or what format should we stole addresses inside the address file

is there a decent way to get current server address without using a cvar

in future we can probably send some more info but that's just a UI problem and doesn't require much structural change besides a new type of message

File: Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesWindow.xaml.cs [hunk 1]
```diff
@@ -59,5 +59,23 @@ public void AddDenied(int denied)
...
+        {
+            var textLabel = new RichTextLabel();
+            textLabel.SetMessage(Loc.GetString("ghost-roles-window-redirect-label"));
+            EntryContainer.AddChild(textLabel);
+
+            var redirectButton = new Button();
+            redirectButton.Text = Loc.GetString("ghost-roles-window-redirect");
+            EntryContainer.AddChild(redirectButton);
+            redirectButton.OnPressed += _ => Redirect();
+        }
```
Comment (Vordenburg) [inline Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesWindow.xaml.cs:74]: Does `enabled` not do anything here for now?

**Comment by Elijahrane** [comment]
so some proposals:
1. the format ss14://example.moe:1212 is the most familiar format, so we'll standardize storing them as that. if there is no port, we'll try appending 1212 as that's the default port.
2. The kick reason passthrough probably needs sanitation and a better separator than newline

**Comment by Elijahrane** [comment]
okay, we can just ask that as part of setting this up people have 
```hub.advertise_url``` set. Seeing our own address is a solved problem.

File: Content.Client/Nyanotrasen/Redial/RedialSystem.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+namespace Content.Client.Redial;
+
+// We can't raise network event outside of an entity system I don't think...
+public sealed class RedialSystem : EntitySystem
+{
+    public void TryRedialToRandom()
+    {
+        RaiseNetworkEvent(new RequestRandomRedialServer());
+    }
+}
```
Comment (Vordenburg) [inline Content.Client/Nyanotrasen/Redial/RedialSystem.cs:13]: Does the following not work from within `Execute` on the `redialrandom` command?

```csharp
IoCManager.Resolve<IEntityManager>().EntityNetManager?.SendSystemNetworkMessage(new RequestRandomRedialServer());
```

File: Content.Client/Launcher/LauncherConnecting.cs [hunk 3]
```diff
@@ -95,6 +120,27 @@ public void RetryConnect()
...
+                    return true;
+                }
+                else
+                {
+                    Logger.InfoS("launcher-ui", $"Redirect not possible, no Ss14Address");
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.ErrorS("launcher-ui", $"Redial exception: {ex}");
```
Comment (Vordenburg) [inline Content.Client/Launcher/LauncherConnecting.cs:140]: I believe the Logger singleton is being phased out eventually.

File: Content.Server/Nyanotrasen/Redial/RedialManager.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    [Dependency] private readonly IConfigurationManager _cfg = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IServerNetManager _net = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+
+    private readonly HttpClient _http = new();
+
+    private TimeSpan _nextUpdateTime = TimeSpan.Zero;
+    private TimeSpan _updateRate = TimeSpan.FromSeconds(30);
+    private List<string> _validServers = new();
```
Comment (Vordenburg) [inline Content.Server/Nyanotrasen/Redial/RedialManager.cs:31]:
```suggestion
    private List<string> _validServers = new();
    
```

File: Content.Server/Nyanotrasen/Redial/RedialServerCommands.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+using Content.Server.Administration;
+using Content.Shared.Administration;
+using Robust.Shared.Console;
+using Robust.Server.Player;
+
+namespace Content.Server.Redial
+{
+    [AdminCommand(AdminFlags.Admin)]
+    public sealed class SendRedialToOtherCommand : IConsoleCommand
```
Comment (Vordenburg) [inline Content.Server/Nyanotrasen/Redial/RedialServerCommands.cs:9]: There's a good example of how to add autocomplete to commands under the KickCommand class that might work well here.

File: Resources/Locale/en-US/launcher/launcher-connecting.ftl [hunk 6]
```diff
@@ -4,6 +4,7 @@ connecting-title = Space Station 14
 connecting-exit = Exit
 connecting-retry = Retry
 connecting-reconnect = Reconnect
+connecting-redirect = Redirect to server with room
```
Comment (Vordenburg) [inline Resources/Locale/en-US/launcher/launcher-connecting.ftl]: I was wondering if one of these might be a better way of wording it:

- Redirect to open server
- Redirect to server with openings
- Redirect to server with availability

File: Content.Shared/Nyanotrasen/Redial/SharedRedial.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+using Lidgren.Network;
+using Robust.Shared.Network;
+
+namespace Content.Shared.Redial
+{
+    /// <summary>
+    ///     A client request for servers to connect to.
+    ///     Should this also be a net message? I dunno it's sending no actual info.
+    /// </summary>
+    [Serializable, NetSerializable]
```
Comment (Vordenburg) [inline Content.Shared/Nyanotrasen/Redial/SharedRedial.cs:12]: I believe any event that is passed through RaiseNetworkEvent requires this attribute.

**Comment by Elijahrane** [comment]
I think I have solved all the issues I brought up earlier

File: Content.Client/Launcher/LauncherConnecting.cs [hunk 8]
```diff
@@ -95,6 +121,27 @@ public void RetryConnect()
...
+            try
+            {
+                if (RedirectAddress != null)
+                {
+                    _gameController.Redial(RedirectAddress);
+                    return true;
+                }
+                else
+                {
+                    Logger.InfoS("launcher-ui", $"Redirect not possible, no Ss14Address");
```
Comment (ZeroDayDaemon) [inline Content.Client/Launcher/LauncherConnecting.cs:135]:
```suggestion
                    Logger.InfoS("launcher-ui", $"Redirect not possible, no SS14Address");
```

File: Resources/Locale/en-US/launcher/launcher-connecting.ftl (hunk 6 above)
Comment (ZeroDayDaemon) [inline Resources/Locale/en-US/launcher/launcher-connecting.ftl]:
```suggestion
connecting-redirect = Connect to another server
```

File: Resources/Locale/en-US/ghost/ghost-gui.ftl [hunk 9]
```diff
@@ -20,3 +20,6 @@ ghost-roles-window-request-role-button-timer = Request ({$time}s)
 ghost-roles-window-follow-role-button = Follow
 ghost-roles-window-no-roles-available-label = There are currently no available ghost roles.
 ghost-roles-window-rules-footer = The button will enable after {$time} seconds (this delay is to make sure you read the rules).
+
+ghost-roles-window-redirect-label = Alternatively, play another server.
+ghost-roles-window-redirect = Redirect
```
Comment (ZeroDayDaemon) [inline Resources/Locale/en-US/ghost/ghost-gui.ftl]:
```suggestion
ghost-roles-window-redirect-label = Alternatively, you can connect to another server.
ghost-roles-window-redirect = Connect to another server
```

**Pull request merged**